src/chopped_graph.hpp

```diff
@@ -0,0 +1,850 @@
+    
+    /// Execute a function on each step of a handle in any path. If it
+    /// returns false, stop iteration. Returns true if we finished and false if
+    /// we stopped early.
+    bool for_each_step_on_handle_impl(const handle_t& handle,
+                                      const std::function<bool(const step_handle_t&)>& iteratee) const;
+    
+    /// For simplicity, we store the same pointer again with the more general type
+    const PathHandleGraph* path_graph;
+    
+    /// This is nontrivial to compute on the fly, so we'll memoize these values
+    spp::sparse_hash_map<path_handle_t, size_t> chopped_steps_count;
+    
+    
+    /// Make some inherited members accesible without scope resolutions
+    using ChoppedGraph<MaxNodeLength>::get_underlying_handle;
+    using ChoppedGraph<MaxNodeLength>::chop_index;
+    using ChoppedGraph<MaxNodeLength>::chopped_size;
+    using ChoppedGraph<MaxNodeLength>::chopped_index_bit_count;
+    using ChoppedGraph<MaxNodeLength>::chopped_index_mask;
+    using ChoppedGraph<MaxNodeLength>::chopped_handle;
+};
+
+/**
+ * The same as ChoppedPathGraph except with the PathPositionHandleGraph interface added
+ */
+template<uint64_t MaxNodeLength = 1024>
+class ChoppedPathPositionGraph : public ChoppedPathGraph<MaxNodeLength>, virtual public PathPositionHandleGraph {
+public:
+    
+    ChoppedPathPositionGraph(const PathPositionHandleGraph& graph);
+    ChoppedPathPositionGraph() = default;
+    virtual ~ChoppedPathPositionGraph() = default;
+    
+    /// Returns the length of a path measured in bases of sequence.
+    size_t get_path_length(const path_handle_t& path_handle) const;
+    
+    /// Returns the position along the path of the beginning of this step measured in
+    /// bases of sequence. In a circular path, positions start at the step returned by
+    /// path_begin().
+    size_t get_position_of_step(const step_handle_t& step) const;
+    
+    /// Returns the step at this position, measured in bases of sequence starting at
+    /// the step returned by path_begin(). If the position is past the end of the
+    /// path, returns path_end().
+    step_handle_t get_step_at_position(const path_handle_t& path, const size_t& position) const;
+    
+protected:
+    
+    const PathPositionHandleGraph* path_position_graph;
+    
+    using ChoppedPathGraph<MaxNodeLength>::chopped_step_handle;
+    using ChoppedPathGraph<MaxNodeLength>::chop_index;
+    using ChoppedPathGraph<MaxNodeLength>::get_underlying_step_handle;
+};
+
+
+
+
+
+
+
+
+/**
+ *
+ * Template implementations
+ *
+ */
+
+template<uint64_t MaxNodeLength>
+ChoppedGraph<MaxNodeLength>::ChoppedGraph(const HandleGraph& graph) : graph(&graph) {
+    
+    // note: assumes dense IDs starting at 1 for efficiency
+    // make a vector where every ID can also index before and after by 1
+    chopped_id_begin.resize(graph.max_node_id() + 2, 0);
+    size_t max_chopped_count = 0;
+    graph.for_each_handle([&](const handle_t& handle) {
```

Comment:
There's no way to avoid this scan of the whole graph on load, is there?

If we added max node length to the stats we expect a graph to track, maybe we could skip it? Then we could also bypass the overlay when the graph isn't in need of chopping, with an OverlayHelper or something.

Reply:
There's no way to avoid it if you want either a) fast ID -> handle queries, or b) fast handle -> sequence/length queries. I have a code path that does something sort of similar over an XG in mpmap though, and it only takes a couple of minutes at start up.